### PR TITLE
Handle `fexists()` in the preprocessor

### LINKED ISSUE #1821: `fexists()` is a constant expression within preprocessor statements
`Error OD0010 at goonstation.dme:30:14: Unexpected identifier fexists in preprocessor expression`

### PR DESCRIPTION
**Pull request by ike709**
Fixes #1821

### CODE REVIEW

File: Content.Tests/DMProject/Tests/Preprocessor/fexists_fail.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// COMPILE ERROR
+#if fexists("fake/path.abc")
+#warn "it somehow exists"
+#else
+#error "it doesn't exist, yay"
```
Comment (ike709) [inline Content.Tests/DMProject/Tests/Preprocessor/fexists_fail.dm:5]:
```suggestion
#if fexists("fake/path.abc")
#error "how could this exist"

```

File: Content.Tests/DMProject/Tests/Preprocessor/fexists_fail.dm [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#if fexists("fake/path.abc")
+#error "how could this exist"
+
```
Comment (ike709) [inline Content.Tests/DMProject/Tests/Preprocessor/fexists_fail.dm:3]:
```suggestion
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 3]
```diff
@@ -314,6 +315,39 @@ internal static class DMPreprocessorParser {
...
+                    Token fexistsInner = Current();
+
+                    if (fexistsInner.Type != TokenType.DM_Resource && fexistsInner.Type != TokenType.DM_ConstantString) {
+                        Error($"Unexpected token {fexistsInner.PrintableText} - resource path expected");
+                        return DegenerateValue;
+                    }
+
+                    Advance();
+                    if (!Check(TokenType.DM_RightParenthesis)) {
+                        DMCompiler.Emit(WarningCode.DefinedMissingParen, token.Location,
```
Comment (ike709) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:334]: Didn't seem worth adding a second warning code.

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 4]
```diff
@@ -314,6 +315,39 @@ internal static class DMPreprocessorParser {
...
+                            "Expected ')' to end fexists() expression");
+                    }
+
+                    Console.WriteLine(Directory.GetCurrentDirectory());
```
Comment (ike709) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:338]:
```suggestion
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 5]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
 
                     return _defines!.ContainsKey(definedInner.Text) ? 1.0f : 0.0f;
                 }
+                else if (token.Text == "fexists") {
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:318]:
```suggestion
                } else if (token.Text == "fexists") {
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 6]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
+                    Advance();
+                    if (!Check(TokenType.DM_LeftParenthesis)) {
+                        Error("Expected '(' to begin fexists() expression");
+                        return DegenerateValue;
+                    }
+
+                    Token fexistsInner = Current();
+
+                    if (fexistsInner.Type != TokenType.DM_Resource && fexistsInner.Type != TokenType.DM_ConstantString) {
+                        Error($"Unexpected token {fexistsInner.PrintableText} - resource path expected");
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:328]: `#if fexists()` doesn't take a resource path as an argument.
```suggestion
                    if (fexistsInner.Type != TokenType.DM_ConstantString) {
                        Error($"Unexpected token {fexistsInner.PrintableText} - file path expected");
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 7]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
                     return _defines!.ContainsKey(definedInner.Text) ? 1.0f : 0.0f;
                 }
+                else if (token.Text == "fexists") {
+                    Advance();
+                    if (!Check(TokenType.DM_LeftParenthesis)) {
+                        Error("Expected '(' to begin fexists() expression");
+                        return DegenerateValue;
+                    }
+
+                    Token fexistsInner = Current();
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:325]: Rename to `fExistsInner`

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 8]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
+                        return DegenerateValue;
+                    }
+
+                    Advance();
+                    if (!Check(TokenType.DM_RightParenthesis)) {
+                        DMCompiler.Emit(WarningCode.DefinedMissingParen, token.Location,
+                            "Expected ')' to end fexists() expression");
+                    }
+
+
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:338]:
```suggestion
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 9]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
+                    var filePath = Path.GetRelativePath(".", fexistsInner.Value.ToString().Replace('\\', '/'));
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:339]: We ensure this is a `DM_ConstantString` token earlier, which should always have a value.
```suggestion
                    var filePath = Path.GetRelativePath(".", fexistsInner.Value!.ToString().Replace('\\', '/'));
```

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 10]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
+
+                    var outputDir = Path.GetDirectoryName(DMCompiler.Settings.Files?[0]) ?? "/";
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:341]: In BYOND `#if fexists()` is relative to the current file and not the root of the project.

File: DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs [hunk 11]
```diff
@@ -314,6 +315,38 @@ internal static class DMPreprocessorParser {
...
+                    if (string.IsNullOrEmpty(outputDir))
+                        outputDir = "./";
+
+                    filePath = Path.Combine(outputDir, filePath);
+
+                    var meep =  File.Exists(filePath) ? 1.0f : 0.0f;
+                    return meep;
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs:348]:
```suggestion
                    return File.Exists(filePath) ? 1.0f : 0.0f;
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**